internal/controller/numaflowcontrollerrollout_controller.go

```diff
@@ -364,6 +374,53 @@ func (r *NumaflowControllerRolloutReconciler) getResourceOperations() (kubeUtil.
 	return ops, cleanup, nil
 }
 
+func (r *NumaflowControllerRolloutReconciler) processNumaflowControllerStatus(ctx context.Context, controllerRollout *apiv1.NumaflowControllerRollout) error {
+	numaLogger := logger.FromContext(ctx)
+
+	// Try to get the Numaflow Controller Deployment
+	deployment := kubernetes.GenericObject{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "numaflow-controller",
+			Namespace: controllerRollout.Namespace,
+		},
+	}
+	existingDeployment, err := kubernetes.GetCR(ctx, r.restConfig, &deployment, "deployments")
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// This could just mean that the deployment has not yet been created.
+			// Log and return to wait for next reconcile call once the deployment is created triggering the reconciliation.
+			numaLogger.Warnf("Numaflow Controller Deployment not found. It may still have to be created.")
+		}
+
+		return fmt.Errorf("error getting the Numaflow Controller Deployment for NumaflowControllerRollout %s/%s: %v", controllerRollout.Namespace, controllerRollout.Name, err)
+	}
+
+	// Parse the status of the existing Numaflow Controller Deployment
+	var existingDeploymentStatus appsv1.DeploymentStatus
+	if len(existingDeployment.Status.Raw) > 0 {
+		err = json.Unmarshal(existingDeployment.Status.Raw, &existingDeploymentStatus)
+		if err != nil {
+			return fmt.Errorf("unable to parse status for existing Numaflow Controller Deployment %s/%s: %v", existingDeployment.Namespace, existingDeployment.Name, err)
+		}
+	}
+
+	if existingDeployment.Generation == existingDeploymentStatus.ObservedGeneration {
+		if existingDeploymentStatus.Replicas == existingDeploymentStatus.ReadyReplicas {
```

Comment:
I think check the complete condition is the right way to tell the deployment is completed successful. https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#complete-deployment

Reply:
From the document:
> The condition holds even when availability of replicas changes (which does instead affect the Available condition).
Should we check that both the `Available` and `Progressing` conditions are set to `True`?

Reply:
Used similar code as that used by ArgoProj gitops-engine as discussed with @xdevxy.